[Alerting][Docs] Adds Alerting & Task Manager Scalability Guidance & Health Monitoring (#91171)
elastic/kibana: docs/user/production-considerations/task-manager-production-considerations.asciidoc

[role="xpack"]
[[task-manager]]
== Task Manager

> {kib} uses background tasks to distribute work across multiple {kib} instances.

Author: 👍

[role="xpack"]
[float]
[[task-manager-production-considerations]]
== Production considerations

[float]
=== Deployment considerations

{es} and {kib} instances use the system clock to determine the current time. To ensure schedules are triggered when expected, you should synchronize the clocks of all nodes in the cluster using a time service such as http://www.ntp.org/[Network Time Protocol].

[float]
[[task-manager-scaling-guidance]]
=== Scaling Guidance

How you deploy {kib} largely depends on your use case. With that in mind, it is difficult to predict the throughout a deplyment might require in order to support {kib} Task Management is difficult, as features can schedule an unpredictable number of tasks at a variety of scheduled cadences.

That said, there is a relatively straight forward method you can follow to produce a rough estimate based on your expected usage.

[float]
[[task-manager-default-scaling]]
==== Default Scale

By default {kib} polls for tasks at a rate of 10 tasks every 3 seconds.
This means that we can reasonably expect a single {kib} instance to support up to 200 _tasks per minute_ (hence forth termed as `200/tpm`).

> In practice, a {kib} instance will only ever achieve that upper bound of `200/tpm` if the duration of task execution is below the polling rate of 3 seconds. For the most part, the duration of tasks is below that threshold, but it can vary greatly as {es} and {kib} usage grow and task complexity increases (such as Alerts executing heavy queries across large datasets).

Author: We shouldn't specifically mention alerts here - this is true for all tasks, not just alerts.
I'll be explicit though that it is a task executing the heavy query...

==== Scaling Horizontaly

At times the sustainable approach might be to expand the throughput of your cluster by provisioning additional {kib} instances.
By default, each additional {kib} instance will add an additional 10 tasks that your cluster can run concurrenctly, but you can also scale each {kib} instance vertically, if your diagnosis indicates they can handle the additional workload.

[float]
[[task-manager-scaling-vertically]]
==== Scaling Vertically

Other times it, might be preferable to increase the throughput of individual {kib} instances.

Tweak the *Max Workers* via the <<task-manager-settings,`xpack.task_manager.max_workers`>> setting, which would allow each {kib} to pull a higher number of tasks per interval, but keep in mind that this could impact the performance of each {kib} instance as _their_ workload would be higher.

Tweak the *Poll Interval* via the <<task-manager-settings,`xpack.task_manager.poll_interval`>> setting, which would allow each {kib} to pull scheduled tasks at a higher rate, but keep in mind that this could impact the performance of the {es} cluster as _their_ workload would be higher.

[float]
[[task-manager-choosing-scaling-strategy]]

> ==== Choosing a Scaling Strategy

Reviewer: ```suggestion
==== Choosing a scaling strategy
```

> Each scaling strategy comes with its own considerations, and as a consequence, the appropriate strategy largely depends on your use case

Reviewer: ```suggestion
Each scaling strategy comes with its own considerations, and the appropriate strategy largely depends on your use case.
```

> Scaling {kib} instances vertically causes higher resource utilization in each {kib} instance, as it will perform more concurrent work.

Reviewer: ```suggestion
Scaling {kib} instances vertically causes higher resource usage in each {kib} instance, as it will perform more concurrent work.
```

> Scaling {kib} instances horizontally requires a higher degree of coordination which can impact overall performance.

Reviewer: ```suggestion
Scaling {kib} instances horizontally requires a higher degree of coordination, which can impact overall performance.
```

> Finding the sweet spot between both strategies requires experimentation, but in general our recommended strategy would be to follow these steps:

Reviewer: ```suggestion
A recommended strategy is to follow these steps:
```

> 1. Produce a <<task-manager-rough-throughput-estimation,rough throughput estimate>> as a guide to provisioning as many {kib} instances as needed. We recommend including any growth in tasks you predict experiencing in the near future, and a buffer to better address ad-hoc tasks.

Reviewer: ```suggestion
1. Produce a <<task-manager-rough-throughput-estimation,rough throughput estimate>> as a guide to provisioning as many {kib} instances as needed. Include any growth in tasks that you predict experiencing in the near future, and a buffer to better address ad-hoc tasks.
```

> 2. After provisioning a deployment as per the rough estimate, assess whether the provisioned {kib} instances achieve the required throughput by evaluating the <<task-manager-health-monitoring>> as described under <<task-manager-theory-insufficient-throughput,"Insufficient Throughtput to Handle the Scheduled Workload">>

Reviewer: ```suggestion
2. After provisioning a deployment, assess whether the provisioned {kib} instances achieve the required throughput by evaluating the <<task-manager-health-monitoring>> as described in <<task-manager-theory-insufficient-throughput, Insufficient throughtput to handle the scheduled workload>>.
```

> 3. If the throughput is insufficient, and {kib} instances are exhibiting low resource utilization, incrementally scale vertically all the while <<kibana-page,monitoring>> the impact of these changes.

Reviewer: ```suggestion
3. If the throughput is insufficient, and {kib} instances exhibit low resource usage, incrementally scale vertically while <<kibana-page,monitoring>> the impact of these changes.
```

> 4. If the throughput is insufficient, and {kib} instances are exhibiting high resource utilization, incrementally scale horizontally by provisioning new {kib} instances and reassess.

Reviewer: ```suggestion
4. If the throughput is insufficient, and {kib} instances are exhibiting high resource usage, incrementally scale horizontally by provisioning new {kib} instances and reassess.
```

> {kib} Task Manager, like the rest of the Elastic Stack, has been designed to scale horizontally, and we recommend taking advantage of this ability to ensure mission ciritcal services such as Alerting and Reporting always have the capacity they need.

Reviewer: ```suggestion
Task Manager, like the rest of the Elastic Stack, is designed to scale horizontally. Take advantage of this ability to ensure mission critical services, such as Alerting and Reporting, always have the capacity they need.
```

> Scaling horizontally requires a higher degree of coordination between {kib} instances. One way by which {kib} Task Manager coordinates with other instances is delaying its polling schedule to avoid conflicting with other instances.

Reviewer: ```suggestion
Scaling horizontally requires a higher degree of coordination between {kib} instances. One way Task Manager coordinates with other instances is by delaying its polling schedule to avoid conflicts with other instances.
```

> By using <<task-manager-health-monitoring>> to evaluate the <<task-manager-health-evaluate-the-runtime,date of the `last_polling_delay`>> across a deployment, we can estimate the frequency at which Task Manager resets its delay mechanism.

Reviewer: ```suggestion
By using <<task-manager-health-monitoring, health monitoring>> to evaluate the <<task-manager-health-evaluate-the-runtime,date of the `last_polling_delay`>> across a deployment, you can estimate the frequency at which Task Manager resets its delay mechanism.
```

> A higher frequency suggests {kib} instances conflict at a high rate, which can be addressed by scaling vertically rather than horizontally, in effect reducing the required coordination.

Reviewer: ```suggestion
A higher frequency suggests {kib} instances conflict at a high rate, which you can address by scaling vertically rather than horizontally, reducing the required coordination.
```

[float]
[[task-manager-rough-throughput-estimation]]

> ==== Rough Throughput Estimation

Reviewer: ```suggestion
==== Rough throughput estimation
```

> Predicting the required throughout a deployment might need to support {kib} Task Management is difficult, as features can schedule an unpredictable number of tasks at a variety of scheduled cadences.

Reviewer: ```suggestion
Predicting the required throughput a deployment might need to support Task Management is difficult, as features can schedule an unpredictable number of tasks at a variety of scheduled cadences.
```

> That said, a rough lower bound can be estimated which is then used as a guide.

Reviewer: ```suggestion
However, a rough lower bound can be estimated, which is then used as a guide.
```

> Throughput is best thought of as a measurements in _tasks per minute_.

Reviewer: ```suggestion
Throughput is best thought of as a measurements in tasks per minute.
```

> As <<task-manager-default-scaling,mentioned above>>, a default {kib} instance can support up to `200/tpm`.

Reviewer: ```suggestion
A default {kib} instance can support up to `200/tpm`.
```

> Given a hypothetical deployment of 100 recurring tasks, estimating the required throughput depends entirely on their scheduled cadence.

Reviewer: ```suggestion
Given a deployment of 100 recurring tasks, estimating the required throughput depends on the scheduled cadence.
```

Suppose we expect to run 50 tasks at a cadence of `10s`, the other 50 at `20m` and in addition, we expect a couple dozen non-recurring tasks every minute.

> A non-recurring task requires a single execution, which means that a single {kib} instance could execute all 100 tasks in less than a minute, utilizing only half of its capacity. As these tasks are only executed once, the {kib} instance will sit idle once all tasks have executed.

Reviewer: ```suggestion
A non-recurring task requires a single execution, which means that a single {kib} instance could execute all 100 tasks in less than a minute, using only half of its capacity. As these tasks are only executed once, the {kib} instance will sit idle once all tasks are executed.
```

> For that reason, we don't tend to include non-recurring tasks in our _tasks per minute_ calculation. Instead, we ensure a _buffer_ included in the final _lower bound_ to incurr the unpredictable cost of ad-hoc non-recurring tasks.

Reviewer: ```suggestion
For that reason, don't include non-recurring tasks in your _tasks per minute_ calculation. Instead, include a buffer in the final _lower bound_ to incur the cost of ad-hoc non-recurring tasks.
```

> A recurring task requires as many executions as its cadence can fit into a minute. This means that a recurring task with a `10s` schedule will require as `6/tpm`, as it will be executed 6 times per minute. On the other hand, a recurring tasks with a `20m` schedule, will only execute 3 times per hour, meaning it would only require a throughput of `0.05/tpm`, a number so small it is difficult to take it into account.

Reviewer: ```suggestion
A recurring task requires as many executions as its cadence can fit in a minute. A recurring task with a `10s` schedule will require `6/tpm`, as it will execute 6 times per minute. A recurring task with a `20m` schedule only executes 3 times per hour and only requires a throughput of `0.05/tpm`, a number so small it that is difficult to take it into account.
```

> For this reason, we recommend grouping tasks by _tasks per minute_ and _tasks per hour_, as demonstrated under <<task-manager-health-evaluate-the-workload,"Evaluate your workload">>, averaging the _per hour_ measurement across all minutes.

Reviewer: ```suggestion
For this reason, we recommend grouping tasks by _tasks per minute_ and _tasks per hour_, as demonstrated in <<task-manager-health-evaluate-the-workload,Evaluate your workload>>, averaging the _per hour_ measurement across all minutes.
```

> Given our predicted workload, we estimate a lower bound throughput of `340/tpm` (`6/tpm` * 50 + `3/tph` * 50 + 20% buffer).

Reviewer: ```suggestion
Given the predicted workload, you can estimate a lower bound throughput of `340/tpm` (`6/tpm` * 50 + `3/tph` * 50 + 20% buffer).
```

> As a default {kib} instance provides us with a throughput of `200/tpm`, a good starting point for our deployment would appear to be to provision 2 {kib} instances. We would then monitor their performance and reassess as the required throughput becomes clearer.

Reviewer: ```suggestion
As a default, a {kib} instance provides a throughput of `200/tpm`. A good starting point for your deployment is to provision 2 {kib} instances. You could then monitor their performance and reassess as the required throughput becomes clearer.
```

> We recognize this is nothing more than a _rough_ estimate, but this rough _tasks per minute_ provides the lower bound that is needed in order to execute tasks on time.

Reviewer: ```suggestion
Although this is a _rough_ estimate, the  _tasks per minute_ provides the lower bound needed to execute tasks on time.
```

> Once you have calculated the rough _tasks per minute_ estimate, we recommend adding a 20% buffer for non-recurring tasks. How much of a buffer is required largely depends on your use case, and as a result we recommend <<task-manager-health-evaluate-the-workload,evaluating your workload>> as it grows to ensure enough of a buffer is provisioned.

Reviewer: ```suggestion
Once you calculate the rough _tasks per minute_ estimate, add a 20% buffer for non-recurring tasks. How much of a buffer is required largely depends on your use case, so <<task-manager-health-evaluate-the-workload,evaluate your workload>> as it grows to ensure enough of a buffer is provisioned.
```
